### PR TITLE
Refactor ReverseMatch signature to allow for iterator reuse

### PR DESCRIPTION
**Pull request by andrewmains12**
This diff applies a similar approach to that found in https://github.com/m3db/m3/pull/3988 to ReverseMatch. The reasoning is the same--ReverseMatch acquires a lot of SortedTagIterators, and pooling is expensive in this code path (contention). Also, the SortedTagIterators are intentionally not closed as of now (see https://github.com/m3db/m3/blob/a873a01a6064a4b5a2538aba4421bb13b8de986c/src/x/serialize/unchecked_metric_tags_iterator.go#L68, which panics if they are), making pooling more or less useless anyhow.

Callers can now pass in a dummy `SortedTagIteratorFn` which actually reuses the same iterator. We may be able to clean up the interfaces further; @prateek has pointed out that the contract of SortedTagIteratorFn might be confusing now--you'd expect to be able to do:

```
iter1 := matchOpts.SortedTagIteratorFn(..)
iter2 := matchOpts.SortedTagIteratorFn(...)

// Use both iter1 and iter2
```

That won't work with iterator reuse, which isn't immediately clear from the function itself. 

Alternatives considered:

- I tried just passing in a SortedTagIterator for reuse directly, but that runs into trouble with the implementation of `uncheckedMetricTagsIter`, which is both a tags iterator *and* an ID. Calling Reset on it with the wrong data messes up the ID bytes as well. Possibly fixable, but I think better to just adopt the same approach for Forward and Reverse match.

### CODE REVIEW

**Comment by andrewmains12** [comment]
Before (with pooling):
image

After (no pooling):

image

File: src/metrics/matcher/config.go [hunk 1]
```diff
@@ -101,21 +100,9 @@ func (cfg *Configuration) NewOptions(
 	}
 
 	// Configure rules options.
-	scope := instrumentOpts.MetricsScope().SubScope("sorted-tag-iterator-pool")
```
Comment (prateek) [inline src/metrics/matcher/config.go:104]: you love to see it

File: src/metrics/filters/tags_filter.go [hunk 2]
```diff
@@ -120,9 +120,6 @@ type TagsFilterOptions struct {
 
 	// Function to extract name and tags from an id.
 	NameAndTagsFn id.NameAndTagsFn
-
-	// Function to create a new sorted tag iterator from id tags.
-	SortedTagIteratorFn id.SortedTagIteratorFn
```
Comment (prateek) [inline src/metrics/filters/tags_filter.go:125]: do you no longer need this because the MatchOpts are passed to the TagFilter?

Comment (andrewmains12) [reply src/metrics/filters/tags_filter.go:125]: Exactly yeah. I think this was already dead code by the time I got here (they were already passed to the TagFilter on the callpath): https://github.com/m3db/m3/blob/4445aa308190c85b517608e80e2c6ce9c82faf91/src/metrics/filters/tags_filter.go#L186

**Review by prateek — APPROVED** [review]
lgtm

correct my understanding: the final change to use this new interface is going to be in the internal repo - where we'll pass down the single iterator per batch by using your trick of returning the same object in the `matchOptions.SortedTagIteratorFn`, yeah?

**Comment by andrewmains12** [comment]
> lgtm
> 
> correct my understanding: the final change to use this new interface is going to be in the internal repo - where we'll pass down the single iterator per batch by using your trick of returning the same object in the `matchOptions.SortedTagIteratorFn`, yeah?

Yep! I'll send you the corresponding diff internally.

**Pull request merged**